File: P3/Python/main.py
import etc
import tabulation
import memoization

flagT = 0
flagF = 0
flagDi = 0
flagDo = 0
flagMem = 0
flagTab = 0

args, flagF, flagT, flagDi, flagDo, flagMem, flagTab = etc.ProcesaArgumentos(flagT, flagF, flagDi, flagDo, flagMem,
                                                                             flagTab)

path = args.file

fichero = etc.abrirFichero(path)
# Opcion de mostrar el contenido del archivo que se pasa por parametro
if flagDi == 1:
    print "File name: " + path + "\n"
    for linea in fichero:
        print linea
    etc.cerrarFichero(fichero)

# Comprobamos errores
if (flagT == 0) and (flagDi == 0) and (flagDo == 0):
    print "Para que el programa funciones correctemente use alguno de los comandos disponibles, 'python main.py -h'"
    exit(1)

# Comprobamos errores
if (flagDo == 0 or flagT == 0) and flagTab == 0 and flagMem == 0 and flagDi == 0:
    print "Para mostrar el output ha de indicar que metodo quiere aplicar, indique -mem o -tab"
    exit(1)

if flagDo == 1 or flagT == 1:

    if flagDi == 1:
        fichero = etc.abrirFichero(path)

    for linea in fichero:

        arr = etc.lstStringToInt(linea)
        size = len(arr)
        # Necesitamos al menos 2 matrices
        if size < 3:
            print "El array no es valido, minimo necesito 3 elementos"
        else:
            # minimo de operaciones y tiempo de tabulation y memoization
            minTab = 0
            minMem = 0
            timeTab = 0
            timeMem = 0
            # Inicializamos el dictionary que vamos a usar para resolver por memoization
            if flagMem == 1:
                memoization.clean()
                memoization.init(arr)

            print "n matrices => " + str(size-1)

            if flagTab == 1:
                # Primera medida de tiempo
                if flagT == 1: etc.time.clock()

                minTab = tabulation.MatrixChainOrder(arr, size)
                # Segunda medida de tiempo
                if flagT == 1: timeTab = etc.time.clock()

            if flagMem == 1:
                # Primera medida de tiempo
                if flagT == 1: etc.time.clock()

                minMem = memoization.MatrixChainOrder(1, size-1)
                # Segunda medida de tiempo
                if flagT == 1: timeMem = etc.time.clock()
            # Creamos el output
            if flagDo == 1:
                if flagTab == 1: print "minimo de operaciones (Tabulation) => " + str(minTab)
                if flagMem == 1: print "minimo de operaciones (Memoization) => " + str(minMem)
            if flagT == 1:
                if flagTab == 1: print "Tiempo (Tabulation) => " + str(timeTab) + " s"
                if flagMem == 1: print "Tiempo (Memoization) => " + str(timeMem) + " s"
        print "--------------------------------------------------------"

etc.cerrarFichero(fichero)

File: P3/C/Main.c
#include <stdio.h>
#include <stdlib.h>
#include <glib.h>
#include <stdbool.h>
#include "Etc.h"
#include "Tabulation.h"
#include "Memoization.h"

GHashTable * map = NULL;
bool flagDi = false;
bool flagDo = false;
bool flagF = false;
bool flagT = false;
bool flagMem = false;
bool flagTab = false;

int main(int argc, char * argv[]) {

	char path[500]="";	// dirección del fichero
	int size;	// Num de elementos
	double timer_tabulation;
	double timer_memoization;

	// Se leen los parámetros pasados al programa
	GetParam(argc, argv, &flagDi, &flagDo, &flagF, &flagT, &flagMem, &flagTab, path);
	int min_operaciones_Tab;
	int min_operaciones_Mem;
	if (flagDi) Input();

	if (!flagDi && !flagDo && !flagT) {
		printf("[!] ERROR. --do || --di || -t requerido\n");
		exit(1);
	}
	if ((!flagT || !flagDo) && !flagTab && !flagMem && !flagDi) {
		printf("[!] ERROR. --tab  || --mem requerido\n");
		exit(1);
	}
	if (flagDo || flagT){

		FILE *fichero = fopen(path, "r");	// Se abre el fichero
		if (fichero != NULL) {
			char linea[50000];	// String que contendrá cada línea del fichero

			while (fgets(linea, sizeof linea, fichero ) != NULL) {

				size = getSizeArray(linea);		// Obtenemos el número de elementos

				if (size < 3) {
					printf(" Aviso: Array no válido. Mínimo 3 elementos \n");
					continue;
				}
				printf("Size => %d\n", (size-1));
				int array[size];
				split(linea, array, size);	// Se crea el array con los elementos

				if (flagTab) {
					if (flagT) ComienzaTimer();
					min_operaciones_Tab = Tabulation(array, size);	// Cáculo de la op min
					if (flagT) timer_tabulation = FinTimer();
				}

				if (flagMem) {
					if (flagT) ComienzaTimer();
					initMemoization(array, size);	// Inicio del hashmap de resultados
					min_operaciones_Mem = MatrixChainOrder(1, size-1);	// Cáculo de la op min
					if (flagT) timer_memoization = FinTimer();
					// g_hash_table_foreach(map, (GHFunc)iterator, "%s -> %d\n");
					clean();	// Destrucción del hashmap creado
				}

				if (flagDo) {
					if (flagTab) printf("[Tabulation] Operaciones mínimas => %d\n", min_operaciones_Tab);
					if (flagMem) printf("[Memoization] Operaciones mínimas => %d\n", min_operaciones_Mem);
				}

				if (flagT) {
					if (flagTab) printf("[Tabulation] Tiempo => %fs \n", timer_tabulation);
					if (flagMem) printf("[Memoization] Tiempo => %fs \n", timer_memoization);
					if ((timer_tabulation > 120.0) || (timer_memoization> 120.0)){

						printf("Max time\n");
						return 0;
					}
				}
				if (flagDo || flagT)printf("-------------------------------------------------\n");
			}
			pclose(fichero);

		} else {

			perror(path);
			return (1);

		}
	}
	return 0;
}
